Lund fixes for 4.1.X

Jakob37 opened the pull request:
<!--
# genomic-medicine-sweden/tomte pull request

Many thanks for contributing to genomic-medicine-sweden/tomte!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/genomic-medicine-sweden/tomte/tree/master/.github/CONTRIBUTING.md)
-->

Drafting a PR with the changes I have made to get Tomte v4.1.0 to run in Lund.

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/genomic-medicine-sweden/tomte/tree/master/.github/CONTRIBUTING.md)
- [ ] Make sure your code lints (`nf-core pipelines lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

Jakob37 committed:
Omit crashing ANNOTATE_SNV row

Jakob37 committed:
Make Peddy sites optional

Jakob37 committed:
Remove sites arg

Jakob37 committed:
No meta for ped

Jakob37 committed:
Re-disable the ANNOTATE_SNV report collect

Jakob37 committed:
Update DNA meta values

Jakob37 committed:
Prefiltering CollectRNASeqMetrics

Jakob37 committed:
Index bam before filtering

Jakob37 committed:
Default addAF to False

Jakob37 committed:
Merge branch 'dev' into v4.1.0

Jakob37 committed:
PR cleanup

Jakob37 commented on subworkflows/local/analyse_transcripts/main.nf:
```diff
@@ -41,7 +41,7 @@ workflow ANALYSE_TRANSCRIPTS {
     ch_bam_ds_bai
         .map { meta, bam, bai ->
-        [ meta.id, meta.single_end, meta.strandedness, meta.sex, meta.vcf, meta.vcf_tbi, meta.dna_id, bam, bai ]
+            [ meta.id, meta.single_end, meta.strandedness, meta.sex, meta.dna_vcf, meta.dna_vcf_tbi, meta.id, bam, bai ]
```
I guess this is a bug? It is defined as `dna_vcf` and `dna_vcf_tbi` over at the `schema_input.json`.

Jakob37 replied:
I reverted this one. I got this error due to having additional columns in my csv (used by our wrapper pipeline). I'll have to rethink the approach there so that Tomte only sees the csv fields it expects.

A clearer error early on would have saved quite some time. A check to add at some point.

Jakob37 committed:
Prefix samtools

fellen31 commented on workflows/tomte.nf:
```diff
@@ -255,7 +255,6 @@ workflow TOMTE {
         }
 
         ch_multiqc_files = ch_multiqc_files.mix(CALL_VARIANTS.out.stats.collect{it[1]}.ifEmpty([]))
-        ch_multiqc_files = ch_multiqc_files.mix(ANNOTATE_SNV.out.report.collect{it[1]}.ifEmpty([]))
```
Looks like this is a duplicate line (we also have it on 254), so it should be fine to just remove here.

jemten commented:
Is the plan to collect all the necessary changes to solve the bugs you've reported in this PR @Jakob37?

Jakob37 commented on conf/modules/bam_qc.config:
```diff
@@ -15,6 +15,14 @@
+    withName: '.*BAM_QC:SAMTOOLS_VIEW' {
+        ext.args = { [
+            "--output-fmt bam",
+            "-f 0x2",
```
As @jemten mentioned over here, this error is due to changes in Picard and does not need to be resolved within the pipeline. So will revert this change.

https://github.com/genomic-medicine-sweden/tomte/issues/277#issuecomment-3466509719

Jakob37 commented on subworkflows/local/bam_qc/main.nf:
```diff
@@ -15,15 +17,21 @@ workflow BAM_QC {
 
+    // Remove read pairs mapping to different chromosomes
+    // Remove non-primary mappings
+    SAMTOOLS_INDEX(ch_bam)
```
Will revert, fixed by fixing Picard

Jakob37 commented:
> Is the plan to collect all the necessary changes to solve the bugs you've reported in this PR @Jakob37?

Not sure what is best. I was thinking to merge any fixes you guys do into this PR (Peddy etc) and resolve the remaining issues here after discussing approaches with you guys.

How do you prefer approaching this?

Jakob37 committed:
Merge

Jakob37 committed:
Revert Peddy and Picard changes

Jakob37 commented:
I reverted the changes related to Peddy and Picard.

I am not so familiar with working with nf-core modules, so would probably be more efficient for one of you to resolve #277

Only small fixes left beyond that.

fellen31 commented:
> I am not so familiar with working with nf-core modules, so would probably be more efficient for one of you to resolve #277

You can modify an nf-core module in the pipeline to change the container version, but it needs to be patched with [`nf-core modules patch`](https://nf-co.re/docs/nf-core-tools/modules/patch), otherwise the linting will complain.

Jakob37 commented:
> > I am not so familiar with working with nf-core modules, so would probably be more efficient for one of you to resolve #277
> 
> You can modify an nf-core module in the pipeline to change the container version, but it needs to be patched with [`nf-core modules patch`](https://nf-co.re/docs/nf-core-tools/modules/patch), otherwise the linting will complain.

OK, I'll give that a go then

Jakob37 committed:
Downgrade picard to 3.3.0

Jakob37 committed:
Patch Picard version

Jakob37 commented:
Our cluster is unavailable until latest Monday next week. Will try our the downgraded Picard version then and come back here after that.

Jakob37 committed:
Revert formatting for modules.json

Jakob37 commented:
I attempted to update the version of the Picard tools. This worked fine when I ran it through.

But I think I need some guidance for the nf-tests.

I am trying to run `nf-test test` on a local server for the `dev` branch (I have pulled latest version), and it crashes due to differing `versions.yml` md5sums.

```
java.lang.RuntimeException: Different Snapshot:
  [                                                                                                     [
      {                                                                                                     {
          "0": [                                                                                                        "0": [
              "GRCh38.primary_assembly.genome.fa:md5,d41d8cd98f00b204e9800998ecf8427e"                              "GRCh38.primary_assembly.genome.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
          ],                                                                                                    ],
          "1": [                                                                                                        "1": [
              "gencode.v46.primary_assembly.annotation.gtf:md5,d41d8cd98f00b204e9800998ecf8427e"                            "gencode.v46.primary_assembly.annotation.gtf:md5,d41d8cd98f00b204e9800998ecf8427e"
          ],                                                                                                    ],
          "2": [                                                                                                        "2": [
              "versions.yml:md5,a809582924251f125611d2c774f1436e"                                          |                "versions.yml:md5,7c9266e8ba129a9315d77759b073022d"
          ],                                                                                                    ],
          "fasta": [                                                                                            "fasta": [
              "GRCh38.primary_assembly.genome.fa:md5,d41d8cd98f00b204e9800998ecf8427e"                              "GRCh38.primary_assembly.genome.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
          ],                                                                                                    ],
          "gtf": [                                                                                              "gtf": [
              "gencode.v46.primary_assembly.annotation.gtf:md5,d41d8cd98f00b204e9800998ecf8427e"                            "gencode.v46.primary_assembly.annotation.gtf:md5,d41d8cd98f00b204e9800998ecf8427e"
          ],                                                                                                    ],
          "versions": [                                                                                         "versions": [
              "versions.yml:md5,a809582924251f125611d2c774f1436e"                                          |                "versions.yml:md5,7c9266e8ba129a9315d77759b073022d"
          ]                                                                                                     ]
      }                                                                                                     }
  ]                                                                                                     ]
```

Here on GitHub I also see (at least) two crashes. I suspect one is related to differing snapshots (?) and the other due to differing run outcomes (which I would need to check).

1. Version mismatches (do I need to create a new snapshot or?)

```
java.lang.RuntimeException: Different Snapshot:
  --- /tmp/expected8420821405306704828jso +++ /tmp/found11434424092296464377json 
  n                                                                              
  @@ -18,7 @@                             @@ +18,7 @@                            
    18         "insert_size\tAll_Reads.fr   18         "insert_size\tAll_Reads.fr
  _count"                                 _count"                                
    19     ],                               19     ],                            
    20     [                                20     [                             
  ! 21         "versions.yml:md5,525ff00c ! 21         "versions.yml:md5,05c28410
  2ac6befe143f16b1d4b0bba6",              db2da889c4c244fd2d01161f",             
  ! 22         "versions.yml:md5,f2ccf683 ! 22         "versions.yml:md5,5d2ce8e7
  968d7657ac00eaf5cfaf4b7a"               1f24ac4d6da4df6278d2a255"              
    23     ]                                23     ]                             
    24 ]                                    24 ]         
```

2. Missing MAE run. I would like to test this as it might be related to my changing of the `dna_vcf` syntax. But would need to get the base tests mentioned above to get running first. MAE runs OK for my own runs.

First diff:

```
  41         },                           41         },                       
-  42         "DROP_CONFIG_RUN_MAE": {                                         
-  43             "drop_config": 2.2,                                          
-  44             "drop": "1.4.0"                                              
-  45         },                                                               
   46         "DROP_FILTER_RESULTS": {     42         "DROP_FILTER_RESULTS": { 
   47             "drop_filter_results"    43             "drop_filter_results"
: 1.5                                   : 1.5                                  
   48         },          
```

So - a bit vague question. But can you help me with some pointers here.

fellen31 commented:
> 1. Version mismatches (do I need to create a new snapshot or?)

Yes, nf-test checks that the module version in the pipeline matches the version (just the md5sum usually in module/subworkflow tests, the actual version in pipeline level tests) of the snapshot. Since you downgraded a module they no longer match, and you need to update the snapshot with `--update-snapshot`. 

>  2. Missing MAE run. I would like to test this as it might be related to my changing of the `dna_vcf` syntax. But would need to get the base tests mentioned above to get running first. MAE runs OK for my own runs.

I'd need to dig into the MAE PR a bit, because the different types of VCFs was a bit confusing to me. 


But at a glance it looks to me like the pipeline was outputting some MAE files before, e.g. `analyse_transcripts/drop/finequagga_mae_top_hits_clinical.tsv`, and after your changes these are no longer output:

<img width="615" height="139" alt="Screenshot 2025-11-05 at 10 52 08" src="https://github.com/user-attachments/assets/60857e41-8e90-4bb0-8bfa-77765a06ac38" />



It also looks like the `DROP_CONFIG_RUN_MAE` now is not run at all:

<img width="690" height="133" alt="Screenshot 2025-11-05 at 10 49 41" src="https://github.com/user-attachments/assets/f055744e-7876-4c96-a4be-9b60ad1f3b17" />

Are they output/run for you locally when you run the same test? You can run nf-test with `--verbose --debug` to get additional information and get paths to dig into the results and work directories.

Jakob37 committed:
Use the real MAE ID

Jakob37 committed:
Merge branch 'dev' of github.com:genomic-medicine-sweden/tomte into jw/4.1.0_jakob_fixes_clean

Jakob37 committed:
Revert all MAE VCF IDs

Jakob37 commented:
OK, I realized the issue. Edit: I was simply confused by executing TOMTE process directly from our wrapper pipeline.

Running with a CSV in the same format as the tests and my (stub) data went through. Looking now at the failing tests, I think it is only due to the Picard downgrade in two modules.

The remaining question to you guys then is about MAE's `addAF`, which I have disabled for it to work in an offline environment. If it is something you want to have in, I'll put that as an optional flag. Let me know what you think.

When done, I'll see if I can update the test snapshots and wrap up the PR.

Jakob37 commented:
A sidenote. I wonder if would make the tests cleaner if omitting the checks of version.yml md5sums. Version diffs are still shown in the diff of the actual version files anyway. What do you think?

Jakob37 committed:
Changelog

fellen31 commented:
> A sidenote. I wonder if would make the tests cleaner if omitting the checks of version.yml md5sums. Version diffs are still shown in the diff of the actual version files anyway. What do you think?

Generally agree that the version.yml md5sums are not great, but that's how it is on the nf-core level as well. I still think we should catch version changes in the subworkflow tests. I think it will be reworked when versions are output as topic channels and is supported by nf-test, see https://github.com/nf-core/modules/issues/5834 and https://github.com/askimed/nf-test/issues/258.

fellen31 commented:
> The remaining question to you guys then is about MAE's `addAF`, which I have disabled for it to work in an offline environment. If it is something you want to have in, I'll put that as an optional flag. Let me know what you think.

I'll let @jemten comment on that!

Jakob37 commented:
> > A sidenote. I wonder if would make the tests cleaner if omitting the checks of version.yml md5sums. Version diffs are still shown in the diff of the actual version files anyway. What do you think?
> 
> Generally agree that the version.yml md5sums are not great, but that's how it is on the nf-core level as well. I still think we should catch version changes in the subworkflow tests. I think it will be reworked when versions are output as topic channels and is supported by nf-test, see [nf-core/modules#5834](https://github.com/nf-core/modules/issues/5834) and [askimed/nf-test#258](https://github.com/askimed/nf-test/issues/258).

OK, I see. Interesting with the version topic channels.

Jakob37 committed:
Update snapshots

Jakob37 commented:
> > The remaining question to you guys then is about MAE's `addAF`, which I have disabled for it to work in an offline environment. If it is something you want to have in, I'll put that as an optional flag. Let me know what you think.
> 
> I'll let @jemten comment on that!

Any thoughts on this @jemten ? I.e. what to do with the `addAF` parameter here, which makes the pipeline access files over the internet and thus crashes in an offline environment.

Here in MAE config. It is also available for the RNA variant calling config (which we currently aren't running):

https://github.com/genomic-medicine-sweden/tomte/blob/6bb851e20cb42e4dde8e31a75b0377fbe7aa0e5a/bin/drop_config.py#L56

I am fine with just turning them off, or alternatively stringing through a boolean from the config.

This is the final point before wrapping this PR, so I am eagerly awaiting your input 😄

jemten commented:
> > > The remaining question to you guys then is about MAE's `addAF`, which I have disabled for it to work in an offline environment. If it is something you want to have in, I'll put that as an optional flag. Let me know what you think.
> > 
> > 
> > I'll let @jemten comment on that!
> 
> Any thoughts on this @jemten ? I.e. what to do with the `addAF` parameter here, which makes the pipeline access files over the internet and thus crashes in an offline environment.
> 
> Here in MAE config. It is also available for the RNA variant calling config (which we currently aren't running):
> 
> https://github.com/genomic-medicine-sweden/tomte/blob/6bb851e20cb42e4dde8e31a75b0377fbe7aa0e5a/bin/drop_config.py#L56
> 
> I am fine with just turning them off, or alternatively stringing through a boolean from the config.
> 
> This is the final point before wrapping this PR, so I am eagerly awaiting your input 😄

Let's put it as parameter and turn it off by default. Fully agree that we need to be able to run this offline. Thanks for all the work you are putting in @Jakob37 🌟

Jakob37 committed:
MAE as parameter

Jakob37 committed:
Fix DROP syntax

Jakob37 committed:
Correct drop_sample_annot script version in snapshot

Jakob37 committed:
More snap files

Jakob37 committed:
Include new skip AF parameter in MAE tests

Jakob37 committed:
Add AF parameter to ANALYSE_TRANSCRIPTS test

Jakob37 committed:
More tests to take the input

Jakob37 committed:
Too many MAE channels. Remove one.

Jakob37 committed:
Version hashes

Jakob37 commented:
I have added a parameter now `--drop_add_af` now, added it to the docs and tests (`nf-test` was new to me, so hope I got everything right there 🤞 )

Tests looks OK and I an hg002 run looked fine as well.

I think this is ready for review now. Are you able to take a look @fellen31 @jemten ?

fellen31 commented on CHANGELOG.md:
```diff
@@ -7,6 +7,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### `Added`
 
+- `drop_add_af` parameter is provided as a optional and default true parameter. If true, DROP will access the internet to retrieve Gnomad frequencies and include in MAE output. [#281](https://github.com/genomic-medicine-sweden/tomte/pull/281)
+
```
We usually have a table with old and new parameters as well (you can find it further down the changelog for previous versions).

fellen31 commented on CHANGELOG.md:
```diff
@@ -17,6 +19,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
+- Remove duplicate report channel, avoiding crash when skipping VEP [#281](https://github.com/genomic-medicine-sweden/tomte/pull/281)
+- Revert Picard version to not crash CollectRNASeqMetrics when reads map to different chromosomes [#281](https://github.com/genomic-medicine-sweden/tomte/pull/281)
+
```
Same for versions changed. Would be great if you could add it at the bottom here.

fellen31 reviewed (commented):
Very nice!

Jakob37 committed:
Add parameters and versions changed tables to changelog

fellen31 commented:
Sorry to be picky, but could you add the NOTE below the tables as well, e.g. 

```
> [!NOTE]
> Parameter has been updated if both old and new parameter information is present.
> Parameter has been added if just thenew parameter information is present.
> Parameter has been removed if new parameter information isn't present.
```

Jakob37 committed:
Changelog cleanup

Jakob37 committed:
Add notes to changelog tables

Jakob37 committed:
Prettier

Jakob37 commented:
Fixed @fellen31 ! No worries with the pickyness, good to get it right 😃 

Let me know if anything else.

When wrapped, could you help me make a Tomte release? This will (hopefully) be the version used internally for completing our validation of Tomte.

jemten commented on docs/parameters.md:
```diff
@@ -84,6 +84,7 @@ Options related to variant calling
 | `drop_padjcutoff_ae` | Adjusted p-value cut-off for DROP Aberrant Expression module | `number` | 0.05 |  |  |
 | `drop_padjcutoff_as` | Adjusted p-value cut-off for DROP Aberrant Splicing module | `number` | 0.1 |  |  |
 | `drop_zscorecutoff` | Z-score cut-off for DROP Aberrant Expression module | `number` | 0 |  |  |
+| `drop_add_af` | Toggle whether DROP should add allele frequencies to output in MAE module | `boolean` | True |  |  |
```
Could add something about this option requires internet access?

Jakob37 replied:
Good point, added

fellen31 replied:
Picky again here, but we'll try to auto-generate the parameters documentation (`nf-core schema docs -o docs/parameters.md`) from `nextflow_schema.json`, could you add it there?

fellen31 replied:
Otherwise it will be missing next time it's generated.

Jakob37 replied:
OK, good to know! I updated there and regenerated the docs.

Jakob37 committed:
Add note about internet access

Jakob37 commented:
Now all the comments are addressed I think. Do you guys have anything more, or is it ready to wrap?

fellen31 commented:
> Now all the comments are addressed I think. Do you guys have anything more, or is it ready to wrap?

Just https://github.com/genomic-medicine-sweden/tomte/pull/281#discussion_r2513106796, otherwise I think everything looks good! Thanks a lot @Jakob37!

Jakob37 committed:
Update nextflow_schema.json and regenerate docs

fellen31 reviewed (approved):
Awesome!

Jakob37 commented:
Thanks for all the help 🙏
